bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/BulkMutation.java

```diff
@@ -416,28 +417,35 @@ public synchronized void run() {
       } finally {
         addCallback(mutateRowsFuture, operationId);
       }
+      setupStalenessChecker();
     }
 
-    /**
-     * A {@link ComplexOperationStalenessHandler} is a method of last resort in cases where somehow the various
-     * listeners failed to get a response, or had some kind of strange issue.
-     */
-    private ComplexOperationStalenessHandler createRetryHandler() {
-      return new ComplexOperationStalenessHandler() {
+    private void setupStalenessChecker() {
+      if (currentRequestManager == null){
+        setRetryComplete();
+        return;
+      }
+      Runnable runnable = new Runnable() {
         @Override
-        public void performRetryIfStale() {
-          synchronized(Batch.this) {
+        public void run() {
+          synchronized (Batch.this) {
             // If the batchId is null, it means that the operation somehow fails partially,
             // cleanup the retry.
             if (currentRequestManager == null || currentRequestManager.isEmpty()) {
               setRetryComplete();
             } else if (currentRequestManager.isStale()) {
               setFailure(
                 io.grpc.Status.UNKNOWN.withDescription("Stale requests.").asRuntimeException());
+            } else {
+              setupStalenessChecker();
             }
           }
         }
       };
+      if (currentRequestManager.lastRpcSentTimeNanos != null) {
+        long delay = currentRequestManager.calculateTimeUntilStaleNanos();
+        stalenessFuture = retryExecutorService.schedule(runnable, delay, TimeUnit.NANOSECONDS);
```

Comment:
Should stalenessFuture be contained in class Batch instead of BulkMutation? It looks like one batch calling setRetryComplete will cancel the latest batch's future, not the batch that created it.